src/Components/Dropdowns/Composed/TypeAheadDropDown.tsx

```diff
@@ -0,0 +1,320 @@
+
+    if (event.keyCode === 40) {
+      // down arrow
+      newIndex = selectIndex + 1
+    } else if (event.keyCode === 38) {
+      // up arrow
+      newIndex = selectIndex - 1
+    }
+
+    const numItems = shownValues.length
+    const newValueWasHighlighted =
+      numItems && newIndex >= 0 && newIndex < numItems
+    if (newValueWasHighlighted) {
+      setSelectIndex(newIndex)
+      return
+    }
+
+    if (event.keyCode === 13) {
+      // return/enter key
+      // lose focus??, reset the selectIndex to -1, & close the menu:
+      //event.target.blur()
+
+      if (numItems && selectIndex >= 0 && selectIndex < numItems) {
+        // they used the arrows; just pressed return
+        doSelection(shownValues[selectIndex], true)
+      } else {
+        // the person could have been typing and pressed return, need to
+        // make sure the value in the input field is real/legal:
+
+        // but:  if the value they typed is LEGAL (in the list/dropdown values), set it;
+        // else: reset to the previous real/legal value:
+        const foundIndex = itemNames.indexOf(typedValue.toLowerCase())
+
+        if (foundIndex >= 0) {
+          // is a real legal value
+          doSelection(items[foundIndex], true)
+        } else {
+          setTypedValueToSelectedName()
+          setMenuOpen(MenuStatus.Closed)
+          setSelectIndex(-1)
+        }
+      }
+    }
+  }
+
+  /**
+   *  if there is a value, get its name, else show the empty string.
+   * only want to show the default name text when there is an item selected.
+   * */
+  const getDisplayName = (item: SelectableItem | null): string => {
+    if (item && item.id) {
+      return getValueWithBackup(item.name, defaultNameText)
+    }
+    return ''
+  }
+
+  const doSelection = (item: SelectableItem | null, closeMenuNow?: boolean) => {
+    setSelectedItem(item)
+    const actualName = getDisplayName(item)
+    setTypedValue(actualName)
+    backupValue.current = actualName
+    setSelectIndex(-1)
+
+    if (closeMenuNow) {
+      setMenuOpen(MenuStatus.Closed)
+    }
+    onSelect(item)
+  }
+
+  const onClickAwayHere = () => {
+    //  reset to the selected value; if the user typed in
+    //  something not allowed it will go back to the last selected value:
+    setTypedValueToSelectedName(backupValue.current)
+  }
+
+  const dropdownStatus = items.length === 0 ? ComponentStatus.Disabled : status
+
+  const placeText =
+    status === ComponentStatus.Loading ? 'Loading...' : placeholderText
+
+  const selectAllTextInInput = (event?: ChangeEvent<HTMLInputElement>) => {
+    if (event) {
+      event.target.select()
+    }
+  }
+
+  // do rendering:
+  const inputComponent = (
+    <Input
+      placeholder={placeText}
+      onChange={filterVals}
+      value={typedValue}
+      onKeyDown={maybeSelectNextItem}
+      testID={`dropdown-input-typeAhead--${testIdSuffix}`}
+      onClear={clear}
+      status={status}
+      onFocus={selectAllTextInInput}
+    />
```

Comment:
this component is only used in one place, any reason why it's stored into a variable?

Reply:
easier to read.

Reply:
easier to read/use (isolated from rest of component for clarity)

Reply:
This is obviously a matter of opinion. But I'd say that storing something into a variable usually implies it'll be used in more than one place.